### PR TITLE
Update leaderboard instances

### PR DESCRIPTION
**Pull request by marnikdenouden**
# Leaderboard sub table
Connect the leaderboard instance table to dynamic data. For this I updated the leaderboard data processing to create a more solid data bundle that can be used by the front end, but also to make the code more structured.

### CODE REVIEW

File: frontend/src/components/leaderboards/Leaderboard.jsx [hunk 1]
```diff
@@ -181,13 +186,60 @@ function Leaderboard({problemData, rowLimit, showPagination}) {
...
+  columns.push(new MetricColumn(problem.scoring_metric));
+  
+  problem.metrics.forEach((metric) => {
+    if (metric.name != problem.scoring_metric.name) {
+      columns.push(new MetricColumn(metric));
+    }
+  });
+
+  columns.push(new LeaderboardColumn("Download solution", 
+    (entry) => { return <div className="download-cell"><i onClick={getDownloadSingleSolutionOnClickHandler(submission, entry)} role="button" className="bi-download"/></div> },
```
Comment (Doodshoofdaapje) [inline frontend/src/components/leaderboards/Leaderboard.jsx:224]: You can not pass paramters to an onclick listener like this. It needs a reference to a method, currently you are calling the function. Therefore the button does not do anything.

File: frontend/src/components/leaderboards/Leaderboard.jsx [hunk 2]
```diff
@@ -181,13 +186,60 @@ function Leaderboard({problemData, rowLimit, showPagination}) {
...
+      columns.push(new MetricColumn(metric));
+    }
+  });
+
+  columns.push(new LeaderboardColumn("Download solution", 
+    (entry) => { return <div className="download-cell"><i onClick={getDownloadSingleSolutionOnClickHandler(submission, entry)} role="button" className="bi-download"/></div> }, 
+    <div className="download-cell">Download<br/>Solution</div>
+  ));
+  columns.push(new LeaderboardColumn("Visualize", 
+    (entry) => { return <div className="download-cell"><i onClick={getHandleVisualizePiOnClickHandler(submission, entry)} role="button" className="bi-eye"/></div> },
```
Comment (Doodshoofdaapje) [inline frontend/src/components/leaderboards/Leaderboard.jsx:228]: You can not pass paramters to an onclick listener like this. It needs a reference to a method, currently you are calling the function. Therefore the button does not do anything.

File: backend/api/data_processing/leaderboard_entry.py [hunk 3]
```diff
@@ -1,42 +1,30 @@
...
+        submission_results = Result.objects.filter(submission=submission)
+
+        # Compute the aggerate results for the submission and each metric of the problem.
         self.results = dict()
-        for metric in results.keys():
-            scores = results[metric]
-            self.results[metric] = sum(scores) / len(scores)
+        for metric in problem.metrics.all():
+            scores = [scores['score'] for scores in submission_results.filter(metric=metric).values('score')]
+            self.results[metric.name] = sum(scores) / len(scores)
```
Comment (Doodshoofdaapje) [inline backend/api/data_processing/leaderboard_entry.py:27]: Something goes wrong here, aggregates seem very random

**Review by Doodshoofdaapje — CHANGES_REQUESTED** [review]
Fix on click listeners and aggregation

File: frontend/src/components/leaderboards/Leaderboard.jsx (hunk 1 above)
Comment (marnikdenouden) [reply frontend/src/components/leaderboards/Leaderboard.jsx:224]: In fact it does work, as I have tested it. It is not the method itself that is the function for the onclick, but rather a function that creates the on click listener using the submission and entry.

File: frontend/src/components/leaderboards/Leaderboard.jsx (hunk 2 above)
Comment (marnikdenouden) [reply frontend/src/components/leaderboards/Leaderboard.jsx:228]: Same story as for the other getOnClickHandler

Comment (marnikdenouden) [reply frontend/src/components/leaderboards/Leaderboard.jsx:228]: Same story for the function that the getOnClickHandler returns, event != e.

File: frontend/src/components/leaderboards/Leaderboard.jsx (hunk 1 above)
Comment (marnikdenouden) [reply frontend/src/components/leaderboards/Leaderboard.jsx:224]: Thanks for noting, the button functionality was in fact broken, but not due to this. The method getDownloadSingleSolutionOnClickHandler is not the method that runs on click, but as the name suggests gets the on click handler function. What was broken and is not fixed in a new commit is that the on click handler has the parameter named 'event' and used the reference 'e'.

**Review by Doodshoofdaapje — APPROVED** [review]
goode

**Pull request merged**